runtime/vm/MHInterpreter.cpp

```diff
@@ -1353,6 +1366,126 @@ VM_MHInterpreter::getArgSlotsBeforePosition(j9object_t argumentTypes, U_32 argPo
+	j9object_t filterHandleType = getMethodHandleMethodType(methodHandle);
+	j9object_t argumentTypes = getMethodTypeArguments(filterHandleType);
+	U_32 filterArgSlots = getMethodTypeArgSlots(filterHandleType);
+	j9object_t argumentIndices = J9VMJAVALANGINVOKEFILTERARGUMENTSWITHCOMBINERHANDLE_ARGUMENTINDICES(_currentThread, methodHandle);
+	U_32 argumentIndicesCount = J9INDEXABLEOBJECT_SIZE(_currentThread, argumentIndices);
+	UDATA *spFirstFilterArgSlot = _currentThread->sp + filterArgSlots;
+
+	/* [... filterHandle args descriptionBytes MethodTypeFrame] */
+	(void)buildMethodTypeFrame(_currentThread, filterHandleType);
+
+	/* Get combinerHandle, combinerHandle.type and combinerHandle.type.argSlots */
+	j9object_t combinerHandle = getCombinerHandleForFilter(methodHandle);
+	j9object_t combinerType = getMethodHandleMethodType(combinerHandle);
+	U_32 combinerArgSlots = getMethodTypeArgSlots(combinerType);
+
+	/* Add combinerHandle to top of stack and adjust sp to make room for combinerHandle.type.argSlots */
+	/* [... filterHandle args descriptionBytes MethodTypeFrame combinerHandle combinerArgs] */
+	_currentThread->sp -= (combinerArgSlots + 1);
+	((j9object_t *)_currentThread->sp)[combinerArgSlots] = combinerHandle;
+	UDATA *spCombinerSlot = _currentThread->sp + combinerArgSlots;
+
+	/* Copy all arguments specified by argumentIndices from filterHandle to the stack slots used by combinerHandler */
+	U_32 arrayIndex = 0;
+	for (arrayIndex = 0; arrayIndex < argumentIndicesCount; arrayIndex++) {
+		U_32 argumentTypeIndex = (U_32)J9JAVAARRAYOFINT_LOAD(_currentThread, argumentIndices, arrayIndex);
+
+		/* Determine the slot index of each argument specified by the index in argumentIndices */
+		U_32 argumentTypeSlots = getArgSlotsBeforePosition(argumentTypes, argumentTypeIndex);
+
+		/* Determine the type of each argument specified by argumentIndices before copying.
+		* Note: long/double type takes 2 slots while other types take 1 slot.
+		*/
+		j9object_t argumentTypeAtIndex =
+			_objectAccessBarrier->inlineIndexableObjectReadObject(
+				_currentThread,
+				argumentTypes,
+				argumentTypeIndex);
+		J9Class *argumentClassAtIndex = J9VM_J9CLASS_FROM_HEAPCLASS(_currentThread, argumentTypeAtIndex);
+
+		if ((_vm->doubleReflectClass == argumentClassAtIndex)
+			|| (_vm->longReflectClass == argumentClassAtIndex)
+		) {
+			spCombinerSlot -= 2;
+			*(U_64*)spCombinerSlot = *(U_64*)(spFirstFilterArgSlot - argumentTypeSlots - 2);
+		} else {
+			spCombinerSlot -= 1;
+			*spCombinerSlot = *(spFirstFilterArgSlot - argumentTypeSlots - 1);
+		}
+	}
+	/* after loop spCombinerSlot is at the top of the combiner arguments */
+	Assert_VM_true(spCombinerSlot == _currentThread->sp);
+
+	/* [... filterHandle args descriptionBytes MethodTypeFrame filterHandle PlaceHolderFrame combinerHandle combinerArgs] */
+	insertPlaceHolderFrame2(combinerArgSlots, methodHandle, J9VMJAVALANGINVOKEMETHODHANDLE_FILTERARGUMENTSWITHCOMBINERPLACEHOLDER_METHOD(_vm));
```

Comment:
Rather than introducing the new helper, write the combiner MH into the stack after making the placeholder frame.
```C
insertPlaceHolderFrame(combinerArgSlots, methodHandle, J9VMJAVALANGINVOKEMETHODHANDLE_FILTERARGUMENTSWITHCOMBINERPLACEHOLDER_METHOD(_vm));
((j9object_t *)_currentThread->sp)[combinerArgSlots] = combinerHandle;
```

Reply:
updated